File: main/subsets/subsets-3.py
class Solution:
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def subsetsTo(stop):
            if stop == 0:
                yield []
                return

            singleton = [nums[stop - 1]]
            for subset in subsetsTo(stop - 1):
                yield subset
                yield subset + singleton

        return list(subsetsTo(len(nums)))

File: main/subsets/subsets-3.cs
public class Solution {
    public IList<IList<int>> Subsets(int[] nums) {
        IEnumerable<int> MakeSubset(int mask) {
            foreach (var element in nums) {
                if ((mask & 1) != 0) yield return element;
                mask >>= 1;
            }
        }

        var powerset = new IList<int>[1 << nums.Length];

        for (var i = 0; i != powerset.Length; ++i)
            powerset[i] = MakeSubset(i).ToList();

        return powerset;
    }
}
